add description and examples for `$id` keyword (#45)
Intelligence-AI/learnjsonschema.com: content/2020-12/core/id.markdown

  - vocabulary: core
    keyword: $vocabulary
---

> The `$id` keyword declares an identifier for the schema and establishes the base URI for resolving other URI references within the schema. This keyword is resolved against the base URI of the overall object. The top level `$id` of a schema, if present, should be absolute.

Reviewer: Now that I think about it, an interesting point we should clarify on a paragraph here is that if the identifier is a URL (i.e. an HTTP one), it is *common* for such URL to respond with the schema if i.e. resolved on a web browser, but that's not mandatory, as the URL is merely an identifier.

> The `$id` keyword declares an identifier for the schema and establishes the base URI for resolving other URI references within the schema. This keyword is resolved against the base URI of the overall object. The top level `$id` of a schema, if present, should be absolute.

Reviewer: On the last point, we could also say that if you are not planning to make your URIs locatable (i.e. actually accessible over the protocol they declare, like HTTP), we advise to use URNs?

> The `$id` keyword declares an identifier for the schema and establishes the base URI for resolving other URI references within the schema. This keyword is resolved against the base URI of the overall object. The top level `$id` of a schema, if present, should be absolute.

Reviewer: Actually, given that URIs are so fundamental to all of this, maybe add a note suggesting the reader to go read the URI RFC to gain a deeper understanding on how resolution works?

## Examples

{{<schema `Declaring an identifier for the schema`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",

>   "$id": "https://example.com/schemas/address.json",

Reviewer: This says the schema is an address, but the example below is just "Hello World!" Maybe put a fake address in the example instead?

  "type": "string"
}
{{</schema>}}

{{<instance-pass `A string is valid`>}}
"Hello World!"
{{</instance-pass>}}

> - _The `$id` keyword declares the identifier `http://example.com/schemas/address.json` for the schema. This identifier can be used to reference the schema from other parts of the document or external documents._

Reviewer: ```suggestion
- _The `$id` keyword declares the URI `http://example.com/schemas/address.json` as the identifier for the schema. This URI serves as the base URI for resolving other URIs within the schema resource._
```

Or something like that?

{{<schema `Nested subschema with relative $id`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/main-schema",
  "type": "object",
  "properties": {
    "name": { "type": "string" },
    "age": { "type": "number" },
    "currentAddress": {

>       "$id": "/address",

Reviewer: ```suggestion
      "$id": "address",
```

Probably simpler?

      "type": "object",
      "properties": {
        "city": { "type": "string" },
        "postalCode": { "type": "number" }
      },

>       "required": ["city", "postalCode"]

Reviewer: ```suggestion
      "required": [ "city", "postalCode" ]
```

Spaces for consistency?

    },
    "permanentAddress": {

>       "$ref": "https://example.com/address"

Reviewer: ```suggestion
      "$ref": "address"
```

Because the top level identifier is already based on `https://example.com`.

Author: I intentionally kept the absolute URI to explicitly indicate where the `$ref` is being resolved.

Reviewer: Alright!

    "postalCode": 67890
  }
}
{{</instance-pass>}}

- The base URI for this schema is `https://example.com/main-schema`. The _address_ subschema has a relative URI `/address`, which resolving against the base URI will result into `https://example.com/address`. Now this URI can be used to reference the  _address_ schema from other parts of the document or external documents.

{{<schema `Nested subschema with absolute $id`>}}
{
  "$schema": "https://json-schema.org/draft/2020-12/schema",
  "$id": "https://example.com/family-info",
  "type": "object",
  "properties": {
    "name": {
      "$id": "https://example.com/name",
      "type": "string"
    },
    "fatherName": { "$ref": "https://example.com/name" },
    "motherName": { "$ref": "https://example.com/name" }
  },
  "required": ["name", "fatherName", "motherName"]
}
{{</schema>}}

{{<instance-pass `An instance with all the required properties is valid`>}}
{
  "name": "John",
  "fatherName": "Peter",
  "motherName": "Julia"
}
{{</instance-pass>}}

- Here, the _name_ subschema has an absolute URI `https://example.com/name`, which can be used to reference the  _name_ schema from other parts of the document or external documents.

> {{<schema  `Schema with URN as value of $id`>}}

Reviewer: ```suggestion
{{<schema `Schema with URN as value of $id`>}}
```

{

>   "$id": "urn:example:vehicle",

Author: I missed adding `$schema` here, I will add it later on

  "type": "object",
  "properties": {
    "car": {
      "$ref": "urn:example:vehicle:car"
    }
  },
  "$defs": {
    "car": {
      "$id": "urn:example:vehicle:car",
      "type": "object",
      "properties": {
        "brand": {"type": "string"},
        "price": {"type": "number" }
      }
    }
  }
}

> {{</schema>}}

Reviewer: The example looks good. Maybe clarify that with URNs there is no such thing as a relative URN. i.e. you always have to fully type them down?

Reviewer: Also, can you add one example using Tag URIs (http://www.taguri.org), as JSON Schema is trying to promote them a bit? Maybe add a short paragraph explaining what they are? You can find examples on the referencing suite: https://github.com/python-jsonschema/referencing-suite/blob/main/tests/json-schema-draft-2020-12/tag-uris.json

Reviewer: The RFC is also a good read: http://www.faqs.org/rfcs/rfc4151.html

Reviewer: (Tag URIs are a *type* of URNs, so might be worth clarifying that)
